http server for distributed inference

alexarmbr opened the pull request:
Here is some info about this PR that I added to the README

## HTTP Inference Server
 The `AsyncGenerationServer` located in `src/para_attn/server/` is an HTTP server implemented using Flask, designed to handle asynchronous generation requests. It manages the creation and monitoring of the worker processes for performing multi-GPU inference. The server takes a python class with a model specific `setup` and `predict` implemented by the user of the server. The `setup` method must call `dist.init_process_group` and initialize the model, and the `predict` method is used to perform the inference. See [here](src/para_attn/server/predictors.py) for an example.

#### HTTP Endpoints
- **`/generate` (POST)**: Accepts a generation request, validates parameters, and ensures all workers are ready. Returns a unique request ID for tracking.
- **`/request_status/<request_id>` (POST)**: Checks the status of a specific generation request using the provided `request_id`. Determines if the request is in flight, completed, or not found.
- **`/cancel` (POST)**: Cancels an ongoing generation request by setting a cancellation event for worker processes.
- **`/server_status` (GET)**: Checks the health of the server and verifies that all worker processes are functioning correctly.

The server can be used either by importing the `AsyncGenerationServer` class and calling its `run` method:
```python
from para_attn.server import AsyncGenerationServer

server = AsyncGenerationServer(model, host, port, num_devices)
server.run()
```

Or by running the `inference-server` command:
```bash
inference-server --model para_attn.server.predictors:FluxPredictor --host 127.0.0.1 --port 5000 --num_devices 2
```
In this later usage, the model path must be a module path that can be imported by `importlib.import_module`.


# Questions for reviewers
- is `src/para_attn/server` the proper place for the server?
- is `tests/server` the proper place for the unit tests?
- anything that can be improved or refactored?

antferdom commented on pyproject.toml:
Should we rename explicitly `inference-server` to `para-attention-server`?

alexarmbr replied:
I think so

chengzeyi replied:
@alexarmbr Sorry to bother, but can it be para-attn server since this is the most used short name?🤣

alexarmbr replied:
yes! It currently is `para-attn-server`

antferdom commented on setup.py:
```diff
@@ -49,7 +49,7 @@ def fetch_requirements():
             "pytest>=7.0.0,<8.0.0",  # https://github.com/pytest-dev/pytest/issues/12273
             "expecttest",
-            #
+            "typer", # requirement for http server
```
I agree, so we let adding server related dependencies as **optional**. We use then `extras_require` in the `setup.py`,

antferdom commented on src/para_attn/server/inference_worker.py:
```diff
@@ -0,0 +1,97 @@
+    try:
+        # these environment variables need to be set in advance of calling dist.init_process_group
+        # which we require to be called by the model setup() method
+        os.environ['MASTER_ADDR'] = 'localhost'
```
if the `/etc/hosts` file or `DNS` configuration is incorrect or missing entries, localhost may not resolve properly. In such misconfigured environments, using 127.0.0.1 explicit IP can bypass resolution problems, so we guarantee no issues.

antferdom commented on src/para_attn/server/inference_worker.py:
```diff
@@ -0,0 +1,97 @@
+                raise Exception(f"worker {rank} cancellation thread is not alive, exiting")
+
+    except Exception as e:
+        print(f"worker {rank} exception: error {e}", file=sys.stderr)
```
See [deepspeed/logger.py](https://github.com/microsoft/DeepSpeed/blob/8ad487254c486163177f1947becf8b35287c6d05/deepspeed/utils/logging.py#L21). [xDiT logger ](https://github.com/xdit-project/xDiT/blob/a4757c9b169f9f10e87875e4171bb57af90e588f/http-service/host.py#L55)prints worker rank, I like this for debugging.

alexarmbr commented on src/para_attn/server/inference_worker.py:
```diff
@@ -0,0 +1,97 @@
+        # 2. self.pipe is assigned to a DiffusersPipeline
+        assert dist.is_initialized()
+        assert model.pipe is not None
+        assert hasattr(model.pipe, '__call__')
```
yes I like this better!

alexarmbr commented:
@chengzeyi @antferdom  Just pushed a commit which addressed some minor changes, renamed some things, added a more detailed README specific to the server in `src/para_attn/server`, and added a dedicated logger with configurable levels as suggested

chengzeyi reviewed (approved):
LGTM!

alexarmbr commented:
@chengzeyi @antferdom lets discuss the `output_path` argument, I agree that it seems a bit strange that the user of the server provides this argument to generate [here](https://github.com/chengzeyi/ParaAttention/pull/21/files#diff-f666af1f5854e6825aa677d2217c9fd0f7de921c55aab85b36b607c81ddab9fcR229). The reason I had the client provide this, is because the server logic is oblivious to the file type that is being generated. All model specific logic is in the [predictor class](https://github.com/chengzeyi/ParaAttention/pull/21/files#diff-6ab04e796a9921c478886dd4fd8a2d46b682dce3031bcd6fc6ccb39a9147542c), and the predictor may be generating a `.png` or `.mp4`, depending on whether an image model or video model is being run. So if there server were to generate the file path and provide it back to the user, which I think ordinarily would make sense, it wouldn't know what the extension is, unless the client tells it. And if the client is telling it (use .mp4 or use .png), I figured why not just provide the whole file path.


Does this make sense? let me know if you have an alternative design that you think would be better!

chengzeyi commented:
@alexarmbr Thanks for clarifying. However I think the server could have a default output format. And we don't need to write the output to the disk. The user could provide an argument to tell the server to specify the output extension or format. But that could be the internal detail of a specific server.

And using a disk output file to indicate whether the request has completed... is indeed not safe! The file can exist on the disk even if the request has not completed... as the predictor could be in the process of creating and writing the file, we can't say that the request has been completed. Though creating a file is atomic but move a file is atomic so we can work around this by using a `create and move` pattern it is still not a good idea, Given the unnecessary complexity and overhead of it. I suggest using an in-memory design with session id to handle it.

alexarmbr commented:
Just pushed another commit to this branch that gets rid of all disk writing during the lifetime of a request. `Predictor`s now return a base64 encoded string containing output data, along with string that encodes the file extension, like [here](https://github.com/alexarmbr/ParaAttention/blob/http-server/src/para_attn/server/predictors.py#L54).

This is communicated from the worker process to the server via a `mp.Queue` which is backed by process shared memory (see [here](https://github.com/alexarmbr/ParaAttention/blob/http-server/src/para_attn/server/inference_worker.py#L105)).

The server then just passes this data/file extension combo directly through to the client [here](https://github.com/alexarmbr/ParaAttention/blob/http-server/src/para_attn/server/server.py#L316).

I like this design a lot better because using the existence of the file to synchronize as in the previous design creates a race condition as explained by @chengzeyi. With this design, the server still does not know or care about the type of data being generated by the predictor (image or video), it is flexible because the predictor can pass any type of file that it likes.

chengzeyi commented:
@alexarmbr LGTM!

antferdom commented on src/para_attn/server/server.py:
```diff
@@ -0,0 +1,371 @@
+        self.host = host
+        self.port = port
+        self.torch_distributed_port = 29500
+        self.world_size = torch.cuda.device_count() if num_devices is None else num_devices
```
Instead of `torch.cuda.device_count()` we should use `dist.get_world_size()`. Is more correct and have formal guarantees that matches `torch.multiprocessing` **world_size**. In the following command, running directly `cli.py` if we don't pass `num_devices` even if we pass the appropriated `torchrun` arguments, these are override by this predicate: 
```shell
python -m torch.distributed.run --nproc_per_node=2 /workspace/src/para_attn/server/cli.py --model para_attn.server.predictors:FluxPredictor --host 127.0.0.1 --port 5000 --num-devices 2 --log-level INFO
```

alexarmbr replied:
good catch @antferdom! I changed this to use `dist.get_world_size()` rather than `torch.cuda.device_count()`

Note that you would not run this cli with `python -m torch.distributed.run`, since the server handles processes internally you start it with `para-attn-server --model para_attn.server.predictors:FluxPredictor --host 127.0.0.1 --port 5000 --num-devices 2`

alexarmbr commented:
@chengzeyi @antferdom perhaps we should close this PR, since it would be redundant to have both this and the `MPDistRunner`?

antferdom commented:
@alexarmbr @chengzeyi I suggest tagging it as **draft**.

chengzeyi commented:
@antferdom I think this server is still very useful and I am OK to get it merged at anytime!
